cmd/ship.go

```diff
@@ -0,0 +1,117 @@
+package cmd
+
+import (
+	"strings"
+
+	"github.com/Originate/git-town/lib/config"
+	"github.com/Originate/git-town/lib/git"
+	"github.com/Originate/git-town/lib/prompt"
+	"github.com/Originate/git-town/lib/steps"
+	"github.com/Originate/git-town/lib/util"
+
+	"github.com/spf13/cobra"
+)
+
+type ShipConfig struct {
+	InitialBranch       string
+	IsTargetBranchLocal bool
+	TargetBranch        string
+}
+
+type ShipFlags struct {
+	Abort         bool
+	CommitMessage string
+	Continue      bool
+	Undo          bool
+}
+
+var shipFlags ShipFlags
+
+var shipCmd = &cobra.Command{
+	Use:   "ship",
+	Short: "Deliver a completed feature branch",
+	Run: func(cmd *cobra.Command, args []string) {
+		steps.Run(steps.RunOptions{
+			CanSkip:              func() bool { return false },
+			Command:              "ship",
+			IsAbort:              shipFlags.Abort,
+			IsContinue:           shipFlags.Continue,
+			IsSkip:               false,
+			IsUndo:               shipFlags.Undo,
+			SkipMessageGenerator: func() string { return "" },
+			StepListGenerator: func() steps.StepList {
+				shipConfig := checkShipPreconditions(args)
+				return getShipStepList(shipConfig)
+			},
+		})
+	},
+	PreRunE: func(cmd *cobra.Command, args []string) error {
+		return validateMaxArgs(args, 1)
+	},
+}
+
+func checkShipPreconditions(args []string) (result ShipConfig) {
+	result.InitialBranch = git.GetCurrentBranchName()
+	if len(args) == 0 {
+		result.TargetBranch = result.InitialBranch
+		git.EnsureDoesNotHaveUncommitedChanges("Did you mean to commit them before shipping?")
```

Comment:
If you supply a branch name that is different then what you are currently on, we stash the changes. I'm actually surprised this is passing. Looks like we are missing a test case where you supply the current branch name